File: HW4/userid/Duck.cpp
#include <iostream>
#include "Duck.h"

Duck::Duck( ) { }

void Duck::swim( ) {
   std::cout << "All ducks float" << std::endl;
}

void Duck::performQuack ( ) {
   quackBehavior->quack( );
}

void Duck::performFly( ) {
   flyBehavior->fly( );
}

void Duck::performEggLaying( ) {
   eggBehavior->laysEgg( );
}

void Duck::setFlyBehavior(FlyBehavior* fb) {
   flyBehavior = fb;
}
   
void Duck::setQuackBehavior(QuackBehavior* qb) {
   quackBehavior = qb;
}

void Duck::setEggBehavior(laysEggs* eb) {
   eggBehavior = eb;
}

File: HW14/huan1588/Duck.java
public abstract class Duck {

   public Duck( ) { }

   public abstract void display( );
   
   public void swim( ) {
      System.out.println("All ducks float");
   }
   
   public void performQuack ( ) {
      quackBehavior.quack( );
   }
   
   public void performFly( ) {
      flyBehavior.fly( );
   }

   public void performEggLaying( ) {
      eggBehavior.laysEgg();
   }
   
   public void setFlyBehavior(FlyBehavior fb) {
      flyBehavior = fb;
   }
      
   public void setQuackBehavior(QuackBehavior qb) {
      quackBehavior = qb;
   }

   public void setEggBehavior(LaysEgg eb) {
      eggBehavior = eb;
   }

   protected QuackBehavior quackBehavior;
   protected FlyBehavior flyBehavior;
   protected LaysEgg eggBehavior;
}

